### ISSUE TITLE
Annotations for Twisted Finger Tutorial

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2281]
|<img alt="sluggo's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| sluggo reported|
|-|-|
|Trac ID|trac#2281|
|Type|enhancement|
|Created|2006-12-07 19:29:52Z|
|Branch|https://github.com/twisted/twisted/tree/finger-annotations-2281|

I've written some annotations for the Twisted Finger Tutorial, which provide background information and explain the unusual Python shortcuts in the examples.  They're at http://sluggo.scrapping.cc/python/twisted_finger_gentle.txt .  Please consider incorporating these into the tutorial.  --Mike Orr <sluggoster@...>

Attachments:

* [introanno.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/02c/02cbcb76348e2688dc6bddf8007fa8b871eafb30/4f988581c7689eae1dd80f0fd6a2dadc024e1c00.diff) (18058 bytes) - added by kehander1 on 2008-10-15 16:40:20Z - Finger Introduction with Mike Orr's annotations integrated
* [protocol.xhtml](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/02c/02cbcb76348e2688dc6bddf8007fa8b871eafb30/0f533f9c257e6bad1f3c8c7e405cda4c1c19b5e1.xhtml) (9950 bytes) - added by kehander1 on 2008-10-15 22:32:11Z - 
* [protocolanno.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/02c/02cbcb76348e2688dc6bddf8007fa8b871eafb30/cd8418b15d94f0d09af43998164f364131a41a52.diff) (9434 bytes) - added by kehander1 on 2008-10-15 22:33:05Z - Finger Part 2 with Mike Orr's annotations integrated
* [introprotoanno.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/02c/02cbcb76348e2688dc6bddf8007fa8b871eafb30/4074890557092b81577aac4e1dfad8c13f2b7197.diff) (21015 bytes) - added by kehander1 on 2008-10-27 23:01:11Z - Intro and Protocol diff (v2)
* [introproto-2281.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/02c/02cbcb76348e2688dc6bddf8007fa8b871eafb30/71477da182e1b7eac22d46b2ccc1f877f600796a.diff) (15534 bytes) - added by kehander1 on 2008-11-28 22:18:28Z - Intro and Protocol diff from 2281

<details><summary>Searchable metadata</summary>

```
trac-id__2281 2281
type__enhancement enhancement
reporter__sluggo sluggo
priority__low low
milestone__ 
branch__branches_finger_annotations_2281 branches/finger-annotations-2281
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1165519792000000 1165519792000000
changetime__1227977266000000 1227977266000000
version__None None
owner__ 
cc__thijs cc__kehander1 cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_1"></a>|@thijstriemstra set owner to @thijstriemstra<br>@thijstriemstra set status to `assigned`|
|-|-|

Moved the suggestion from the document called '17 - Announce on the Web, Too' to [#2110](https://github.com/twisted/twisted/issues/2110)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_2"></a>|@thijstriemstra commented|
|-|-|

Also moved '14 - Setting Message By Local Users' to [#2210](https://github.com/twisted/twisted/issues/2210)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_3"></a>|@thijstriemstra commented|
|-|-|

Replying to [thijs](#note_2):
> Also moved '14 - Setting Message By Local Users' to [#2210](https://github.com/twisted/twisted/issues/2210)
Err, [#2110](https://github.com/twisted/twisted/issues/2110)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_4"></a>|@thijstriemstra commented|
|-|-|

And moved '8 - Output from Non-empty Factory' to [#2110](https://github.com/twisted/twisted/issues/2110)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_5"></a>|@thijstriemstra commented|
|-|-|

Marked [#4924](https://github.com/twisted/twisted/issues/4924) as a duplicate of this one.

**Comment by twisted-trac** [comment]
|<img alt="kehander1's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|kehander1 commented|
|-|-|

I don't know if the <pre class="python"> tag will keep the html from being parsed when it's exported; it doesn't seem to work that way when I open the xhtml in Firefox.

Also, I wanted to add more to "Announce on IRC", but I couldn't quite figure out how the private message was supposed to work.  Maybe it's fine as-is.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to kehander1<br>@exarkun set status to `new`|
|-|-|

< and > inside a pre tag need to be quoted - &lt; and &gt;

**Comment by twisted-trac** [comment]
|<img alt="kehander1's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_11"></a>|kehander1 commented|
|-|-|

But of course.  (You can ignore that xhtml attachment; I just selected that one by mistake.)

If these look halfway-reasonable I can try to do something with components.xhtml, but since Mr. Orr's annotations end here, perhaps that's best left filed under some other ticket.

**Comment by twisted-trac** [comment]
|<img alt="sluggo's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|sluggo commented|
|-|-|

I'm no longer involved in Twisted, so if you could change the "Reported by" or move to a new ticket, I'd appreciate not getting the status change messages.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_13"></a>|@thijstriemstra removed owner|
|-|-|

Replying to [sluggo](#note_12):
> I'm no longer involved in Twisted, so if you could change the "Reported by" or move to a new ticket, I'd appreciate not getting the status change messages.

One solution for that problem is to login to Trac, go to 'User Configuration' on the top and remove your emailaddress there. That should stop those emails I think.

**Comment by twisted-trac** [comment]
|<img alt="sluggo's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|sluggo commented|
|-|-|

OK done.  If I don't comment again that means I'm not getting the emails.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

(In [25000]) Branching to 'finger-annotations-2281'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun commented|
|-|-|

(In [25001]) apply introanno.diff and protocolanno.diff

refs [#2281](https://github.com/twisted/twisted/issues/2281)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun set owner to @thijstriemstra|
|-|-|

1. In the introduction:
    1. The sample finger output should use the shell `/bin/sh`, I think.  That is the standard shell, after all. :P
    1. The parenthetical explaining what a ''daemon'' is should be replaced with a glossary reference.
    1. The parenthetical explaining what happens if you run `finger nail` should be unparenthesized.  It's just more explanation.
    1. ''insteance'' is misspelled.  I'd just drop that word, and the use of ''class'' later in the same paragraph.
    1. The documentation after the `finger02.py` listing about where protocol instances come from is wrong.  The reactor calls the factory's `buildProtocol` method.  The rest is up to the factory.  How to present this without confusing the user may be tricky, but I don't think there's any value in explaining it _wrong_.
    1. Objects are ''garbage collected'', not ''garbage-collected''.  Heaps are ''garbage-collected'' if the objects allocated in them are ''garbage collected''. :)
    1. It looks like the paragraph after the `finger03.py` listing is only reflowed; no content seems to be different.  This just adds noise to the diff and makes the review harder, reverting it might be sensible.
    1. Same for the paragraph after the `finger04.py` listing.
    1. instead of ''a remote Oracle?'' in the paragraph after the `finger08.py` listing, I'd say _a remote Oracle server?_
    1. Quoting the word "threads" in the new paragraph about being friendly to the CPU is pointless.
    1.  The new text about CPU usage and Deferreds is a bit confused.  Deferreds don't give you any of the features this text attributes to them.  They're just a way to keep track of callbacks.  It's everything '''else''' you do in a Twisted program that keeps the CPU busy and lets you not use threads.  As an example of how this text is misleading, Twisted gives you an interface to using threads which involves Deferreds!
    1. Woah.  A huge wall of stuff about Deferreds.  This doesn't really belong here.  If the user needs some hand-holding to understand the introduction of Deferreds in `finger09.py`, that's fine - but it needs to be text integrated with the example code, not a mini-howto that's a page long and introduces new example code and weird new language ("slow functions").  I suggest removing all of this Deferred stuff, including the existing Deferred stuff, and delaying the introduction of Deferreds until the example actually _benfits_ from them.  Tossing complexity "for the hell of it" (quoting the tutorial) is stupid.
  1. In the protocol document
    1. The changes here start off by adding two new snippets of code which are supposed to be equivalent to the code in `finger12.tac`.  This is silly.  If the new code is more readable (it seems to be), then the listing should be updated.  Having two versions is complex, confusing, and verbose.
    1. Ahh, so many changes.  Most of the `protocol.xhtml` changes still need to be reviewed.  There's plenty of feedback to respond to here now though, so progress can be made on this ticket.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun commented|
|-|-|

Well, I dunno how that markup managed to get screwed up, but it did, oh well.

Not sure who the right owner for this ticket is, whoever wants to work on it, feel free to grab it.

**Comment by twisted-trac** [comment]
|<img alt="kehander1's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_19"></a>|kehander1 removed owner|
|-|-|

I agree about the stuff about deferreds.  (I copied it all from Mr. Orr's document mostly for the sake of completeness.)  However, removing all the existing deferred stuff from the examples strikes me as a rather radical re-tooling of the whole tutorial, and I think it would be a good idea to have something at least slightly better in the interim.

This new diff doesn't include any extra code snippets (and was made with branches/finger-python-idioms-2110 in mind).

To address some specific items:
2. How exactly does one add a "glossary reference" ?
7. This paragraph changes "protocol" to "standard finger protocol", which I think makes a lot more sense.
8. This paragraph provides a brief word about what exactly a honeypot is.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun set owner to kehander1|
|-|-|

To add something to the glossary, add a new entry to doc/core/howto/glossary.xhtml.  Then, to link to it, add an `<a href="glossary.xhtml#Name">...</a>`.

It looks like point 11 from the previous review hasn't been addressed.  There's still a misleading paragraph about how Deferreds let you "use the CPU more".  The whole paragraph is really about event loops.  Deferreds are just for keeping track of callbacks.

In the _Do Nothing_ section of `intro.xhtml`, the paragraph which used to talk about how a new protocol instance is created for every connection no longer explains this.  Why was it removed?

Also, the diff _seems_ to be against trunk.  It'd be better if subsequent diffs were against the branch.

**Comment by twisted-trac** [comment]
|<img alt="kehander1's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_21"></a>|kehander1 removed owner|
|-|-|

1.Done.
 1. All right, I took out a little more.
 1. This is in accordance with point #5 of your previous review!
 1. Oops; this is now diffed against the branch (I hope).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_22"></a>|@exarkun commented|
|-|-|

(In [25519]) Apply introproto-2281.diff

refs [#2281](https://github.com/twisted/twisted/issues/2281)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun set owner to @exarkun|
|-|-|

  * some mistakes in the glossary addition - the a tag's name is _Deferred_ which seems unrelated, and there's probably no reasonable API to link to for the term _Daemon_.
  * protocol.xhtml has some ill formed markup, a missing `<p>` or `</p>` in several places

Some of the other language needs to be cleaned up as well.  It's hard to describe how without actually making the changes myself, so I'm going to do that.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun commented|
|-|-|

(In [25520]) fix the daemon entry; add a tac entry so the finger tutorial can link to that too

refs [#2281](https://github.com/twisted/twisted/issues/2281)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_25"></a>|@exarkun commented|
|-|-|

(In [25521]) whitespace cleanup; fix some misleading language

refs [#2281](https://github.com/twisted/twisted/issues/2281)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_26"></a>|@exarkun commented|
|-|-|

(In [25522]) whitespace cleanup

refs [#2281](https://github.com/twisted/twisted/issues/2281)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_27"></a>|@exarkun set status to `closed`|
|-|-|

(In [25523]) Merge finger-annotations-2281

Author: Mike Orr, kehander1
Reviewer: exarkun
Fixes: [#2281](https://github.com/twisted/twisted/issues/2281)

Add numerous clarifications to the finger tutorial.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
